sklearn/preprocessing/tests/test_label.py

```diff
@@ -673,6 +745,57 @@ def test_invalid_input_label_binarize():
         label_binarize([[1, 3]], classes=[1, 2, 3])
 
 
+@pytest.mark.parametrize(
+    "string_labels, y, classes, expected",
+    [
+        [True, [1, 0, 0, 1], ["yes", "no"], [[0], [0], [0], [0]]],
```

Comment:
I don't understand in which context one would expect integer valued `y` and `str` valued `classes`. To me, if `classes` holds `str` values, so would `y`, and it would be fine to raise a `ValueError` if that's not the case.

It seems like a bug not to do so.

If we fix this bug, then, adding array API support should become cleaner: there would be no need to catch the `cupy.asarray(["yes", "no"])` exceptions because they can never be raised.